File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression.R
# Importing the dataset
dataset = read.csv("Position_Salaries.csv")
dataset = dataset[2:3]

# Splitting the dataset in train set and test set
# o dataset é muito pequeno e precisamos de toda informação para fazer as melhores previsões possiveis
# por isso não haverá split
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Profit, SplitRatio = 0.8)
# train_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature scaling
# a biblioteca aplica feature scalling internamente, assim como no python
# train_set$Country = as.numeric(train_set$Country)
# test_set$Country = as.numeric(test_set$Country)
# train_set[,1:3] = scale(train_set[,1:3])
# test_set[,1:3]  = scale(test_set[,1:3])

# Polynomial Linear Regression

# Fitting Linear Regression to the dataset just for comparing
lin_reg = lm(formula = Salary ~ .,
             data = dataset)
summary(lin_reg)

# Fitting Polynomial Regression to the dataset
dataset$Level2 = dataset$Level^2
poly_reg = lm(formula = Salary ~ .,
             data = dataset)
summary(poly_reg)
      
# Visualising the Linear Regression results 
# é passado o dataset como newdata no predict linear. Mesmo o dataset tendo sido alterado (adicionado colunas),
# essas colunas vão ser ignoradas, pois não estavam presentes no momento da criação do lin_reg.
# isso pode ser visto no: summary(lin_reg)
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
            colour = "red") +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = "blue") +
  geom_point(aes(x = 6.5, y = predict(lin_reg, newdata = data.frame(Level = 6.5))),
            colour = "green") +
  ggtitle("Simple Linear Regression") +
  ylab("Level") +
  xlab("Salary")

# Visualising the Polynomial Regression results (degree = 2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = "blue") +
  ggtitle("Polynomial Regression (degree = 2)") +
  ylab("Level") +
  xlab("Salary")

# Visualising the Polynomial Regression results (degree = 4)
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ .,
              data = dataset)
summary(poly_reg)

ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = "blue") +
  
  ggtitle("Polynomial Regression (degree = 4)") +
  ylab("Level") +
  xlab("Salary")

# Visualising the Polynomial Regression results (degree = 4 + grid)
grid = seq(1, 10, by=0.1)
X_grid = data.frame("Level" = grid,
                    "Level2" = grid^2,
                    "Level3" = grid^3,
                    "Level4" = grid^4)

poly_reg = lm(formula = Salary ~ .,
              data = dataset)
summary(poly_reg)

ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +
  geom_line(aes(x = X_grid$Level, y = predict(poly_reg, newdata = X_grid)),
            colour = "blue") +
  geom_point(aes(x = 6.5, y = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                                     Level2 = 6.5^2,
                                                                     Level3 = 6.5^3,
                                                                     Level4 = 6.5^4))),
             colour = "green") +
  ggtitle("Polynomial Regression (degree = 4 + grid)") +
  ylab("Level") +
  xlab("Salary")

# Prediction a new reuslt with Linear Regression
pred_lin = predict(lin_reg, newdata = data.frame(Level = 6.5))

# Prediction a new reuslt with Polyminal Regression (degree = 4 + grid)
pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                  Level2 = 6.5^2,
                                                  Level3 = 6.5^3,
                                                  Level4 = 6.5^4))

File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv("Position_Salaries.csv")
X = dataset.iloc[:, 1:-1].values
Y = dataset.iloc[:, -1].values

# Splitting the dataset in train set and test set
# o dataset é muito pequeno e precisamos de toda informação para fazer as melhores previsões possiveis
# por isso não haverá split
"""
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)
"""

# Feature scaling
# a biblioteca aplica feature scalling internamente automaticamente
"""
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
# IMPORTANTE: aqui faz só o transform, pois o fit já foi feito no train, isso faz com que ambos esteja escaladas na mesma base
X_test = sc_X.transform(X_test)
# em classificações o Y não precisa ser escalado. Em regreção sim
"""

# Polynomial Regression

# Fitting Linear Regression to the dataset just for comparing
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, Y)

# Fitting Polynomial Regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
# se você por o degree 3 vão criadas coulas 2 e 3
# essa função cria a coluna de x0 = 1 para o algoritimo não ignorar b0, igual na Multiple Linear Regression
poly_reg = PolynomialFeatures(degree = 2)
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, Y)

# Visualising the Linear Regression results 
plt.scatter(X,Y, color = "red")
plt.plot(X, lin_reg.predict(X), color = "blue")
plt.scatter(6.5, lin_reg.predict(np.array(6.5).reshape(-1,1)), color = "green")
plt.title("Simple Linear Regression")
plt.xlabel("Position level")
plt.ylabel("Salary")
plt.show()

# Visualising the Polynomial Regression results (degree = 2)
plt.scatter(X,Y, color = "red")
plt.plot(X, lin_reg_2.predict(poly_reg.fit_transform(X)), color = "blue")
plt.title("Polynomial Regression (degree = 2)")
plt.xlabel("Position level")
plt.ylabel("Salary")
plt.show()

# Visualising the Polynomial Regression results (degree = 4)
# quando maior o degree mais preciso a previsão é, porém pode haver overfitting
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, Y)

plt.scatter(X,Y, color = "red")
plt.plot(X, lin_reg_2.predict(poly_reg.fit_transform(X)), color = "blue")
plt.title("Polynomial Regression (degree = 4)")
plt.xlabel("Position level")
plt.ylabel("Salary")
plt.show()

# Visualising the Polynomial Regression results (degree = 4 + grid)
# plotar um grid com mais pontos traz uma melhor suavização ao gráfico
X_grid = np.arange(min(X),max(X), step = 0.1)# vetor
X_grid = X_grid.reshape((len(X_grid),1))# matriz

poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, Y)

plt.scatter(X,Y, color = "red")
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = "blue")
plt.scatter(6.5, lin_reg_2.predict(poly_reg.fit_transform(np.array(6.5).reshape(-1,1))), color = "green")
plt.title("Polynomial Regression (degree = 4) + (grid = 0.1)")
plt.xlabel("Position level")
plt.ylabel("Salary")
plt.show()

# Prediction a new reuslt with Linear Regression
pred_lin = lin_reg.predict(np.array(6.5).reshape(-1,1))

# Prediction a new reuslt with Polyminal Regression (degree = 4 + grid)
pred2_pol = lin_reg_2.predict(poly_reg.fit_transform(np.array(6.5).reshape(-1,1)))
